TST: Use pytest-run-parallel against free-threaded CI

andfoy opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
See https://github.com/scipy/scipy/issues/20669

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR adds an additional CI job on top of the already-existing 3.13 free threaded one, this new CI pipeline runs the SciPy testsuite using the [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel/) plugin, which allows to run individual tests concurrently during a pytest session

#### Additional information
<!--Any additional information you think is important.-->

Several tests were using the global random number generator, via `np.random.seed`, which was causing numeric result mismatches when the tests were being run in parallel. Instead, all tests now use `np.random.RandomState`, which ensures that each thread gets its own independent random generator.

andfoy committed:
TST: Use pytest-run-parallel against free-threaded CI

andfoy committed:
FIX: Further remove global state from tests in cluster and differentiate

andfoy committed:
FIX: Use per-thread random generator in fft tests

andfoy committed:
ENH: Extend uarray local context to consider threading

andfoy committed:
FIX: scipy.fft tests are  now passing using multiple threads

andfoy committed:
FIX: scipy.interpolate and parts  of interpolate are passing under thread-stress

andfoy committed:
ENH: scipy.interpolate tests are now passing under concurrent loads

andfoy committed:
FIX: scipy.io tests now work under concurrent loads

andfoy committed:
FIX: scipy.linalg tests are  now running under concurrent loads

andfoy committed:
FIX: scipy.misc tests are nnow passing under concurrent loads

andfoy committed:
FIX: scipy.ndimage tests are now passing under concurrent loads

andfoy committed:
FIX: scipy.odr tests are now passing under concurrent loads

rgommers reviewed (commented):
Thanks @andfoy, this looks quite good! I did a cursory review, a few thoughts:

- The testing approach seems to find a lot of issues, so that's good. We should add this use of `pytest-run-parallel` in a single CI job indeed.
- I expect that the pytest mark used needs to be registered in `conftest.py` (similar to the `slow` mark) in order for the test suite not to fail when the plugin isn't installed.
- Logistics: let's aim to resolve the merge conflicts and get this merged. Better to do the second half of submodules in a second PR, because this is already huge and reviewing it in detail is a large enough job as is.
- The changes seem to consist of three types, all of which are useful:
    - locking around calls to non-reentrant compiled extensions
    - fixing or protecting usage of global state in pure Python code
    - fixes to the test suite to enable it to run in parallel under `threading`

andfoy committed:
FIX: scipy.optimize tests are now passing under concurrency

andfoy committed:
FIX: scipy.signal tests are passing under concurrent loads

andfoy committed:
ENH: Define parallel_threads marker stub in case pytest-run-parallel is not available

andfoy committed:
FIX: Solve leftover merge regions

andfoy committed:
FIX: Add stub num_parallel_threads fixture

andfoy committed:
FIX: Use threading.local in  fft backend tests

andfoy committed:
TST: Mark bspline tests that check for warnings as single-threaded

andfoy committed:
CI: Reduce number of parallel workers

andfoy committed:
FIX: prevent overlapping aaccess to BSpline knots when calling derivative concurrently

andfoy committed:
FIX: Additional fixes in BSppline

andfoy committed:
STY: Address style issues

andfoy committed:
CI: Temporarily disable parallel testing

rgommers commented on .github/workflows/linux.yml:
```diff
@@ -450,10 +450,14 @@ jobs:
     if: >
       needs.get_commit_message.outputs.message == 1
       && (github.repository == 'scipy/scipy' || github.repository == '')
+      && false
```
This disables the jobs completely. I think it's only needed to disable or remove the `parallel: 1` case? `parallel: 0` was already there and passes.

rgommers commented on scipy/integrate/_ivp/tests/test_ivp.py:
```diff
@@ -758,7 +776,10 @@ def early_event(t, y):
-def test_event_dense_output_LSODA():
+def test_event_dense_output_LSODA(num_parallel_threads):
+    if num_parallel_threads > 1:
+        pytest.skip('LSODA does not allow for concurrent execution')
```
Can you make a separate issue for this (assuming we want to fix this)? Is the difference with other solvers that there wasn't a good place to add a lock?

andfoy replied:
It seems that  `lsoda`, `zvode` and `vode` (defined as FORTRAN routines) are not thread-safe, I don't know much about the wrapper generation and how to add locks to their calls, so I deferred the locking to the testing instead.

rgommers replied:
I had a closer look. There is a Python interface that calls the Fortran routines (e.g., `_lsoda.lsoda`). When I remove a test skip I get a crash, that is fixed by putting a lock around that call - here is an example diff:

```diff
diff --git a/scipy/integrate/_ode.py b/scipy/integrate/_ode.py
index 2d52e093e..7dec2dbf5 100644
--- a/scipy/integrate/_ode.py
+++ b/scipy/integrate/_ode.py
@@ -82,6 +82,7 @@ __all__ = ['ode', 'complex_ode']
 
 import re
 import warnings
+from threading import Lock
 
 from numpy import asarray, array, zeros, isscalar, real, imag, vstack
 
@@ -95,6 +96,9 @@ _vode_int_dtype = _vode.types.intvar.dtype
 _lsoda_int_dtype = _lsoda.types.intvar.dtype
 
 
+ODE_LOCK = Lock()
+
+
 # ------------------------------------------------------------------------------
 # User interface
 # ------------------------------------------------------------------------------
@@ -1343,7 +1347,8 @@ class lsoda(IntegratorBase):
             self.acquire_new_handle()
         args = [f, y0, t0, t1] + self.call_args[:-1] + \
                [jac, self.call_args[-1], f_params, 0, jac_params]
-        y1, t, istate = self.runner(*args)
+        with ODE_LOCK:
+            y1, t, istate = self.runner(*args)
         self.istate = istate
         if istate < 0:
             unexpected_istate_msg = f'Unexpected istate={istate:d}'
diff --git a/scipy/integrate/tests/test_integrate.py b/scipy/integrate/tests/test_integrate.py
index f83876d49..0f45b6905 100644
--- a/scipy/integrate/tests/test_integrate.py
+++ b/scipy/integrate/tests/test_integrate.py
@@ -112,11 +112,8 @@ class TestOde(TestODEClass):
                 self._do_problem(problem, 'zvode', 'adams')
             self._do_problem(problem, 'zvode', 'bdf')
 
-    @pytest.mark.thread_unsafe
-    def test_lsoda(self, num_parallel_threads):
+    def test_lsoda(self):
         # Check the lsoda solver
-        if num_parallel_threads > 1:
-            pytest.skip(reason='LSODA does not allow for concurrent execution')
         for problem_cls in PROBLEMS:
             problem = problem_cls()
             if problem.cmplx:
```

The `ode` and `odeint` classes are annoying to read, and the existing `check_handle`/`acquire_new_handle` seems wrong, but a few well-places locks should fix this. Just skipping the tests isn't useful, so let's try to remove those skips again and fix the actual issues.

rgommers replied:
@andfoy this wasn't addressed - I pushed a commit to fix this up. Everything except `integrate.solve_ivp` is now thread-safe.

rgommers commented on scipy/interpolate/_fitpack_impl.py:
```diff
@@ -745,6 +745,7 @@ def splder(tck, n=1):
         return splantider(tck, -n)
 
     t, c, k = tck
+    c = c.copy()
```
`c` is reassigned a few lines below (`c = (c[1:-1-k]  ...`), and not modified - why is this needed?

andfoy replied:
For some reason, without the copy, values in `c` were mutating even if the function definition was not performing any major operation that caused changes in the values stored.

andfoy replied:
There was a concurrency situation on the corresponding tests, I've just fixed them and removed these copies.

rgommers reviewed (commented):
Thanks @andfoy. I went through the whole diff now. Overall it looks good. A few high-level questions about the patterns used:
 
A lot of `np.random.seed(1234)` calls were replaced with `rng = np.random.RandomState(1234)`. Can you explain in the PR description why the latter is safe and the former is not?

Most of the `@pytest.mark.parallel_threads(1)` indicate places where functionality isn't thread-safe apparently. What was the strategy for choosing to add `parallel_threads`? Are those places that are inhererently not fixable, or too much work right now, or ...?

andfoy committed:
FIX: address review comments

rgommers commented:
> 1. Most of the markers were placed in tests that use `pytest.warns`. Since `warnings` is not thread-safe, such tests were bound to fail consistently, adding the markers allow to test for warnings by making sure that the tests are run in a single thread.

In another discussion I saw the suggestion of adding a `thread_unsafe` marker to make that clearer. I'd also be curious if `pytest-run-parallel` has access to the test function's implementation and can therefore introspect it for usage of `pytest.warns` to automatically do the right thing? 

> 2. Other cases that imply the usage of the marker are functions that are not thread-safe by design

It feels like a `reason=` keyword could be helpful here - something to consider for `pytest-run-parallel`?

andfoy commented:
> adding a thread_unsafe marker to make that clearer

We could add `thread_unsafe` an alias of `parallel_threads(1)`.

> introspect it for usage of pytest.warns to automatically do the right thing?

We could do that preemptively for instances of functions that use `pytest.warns` indeed, however, we should also look up for instances of `warnings.catch_warnings`, also, I don't know if `np.testing.suppress_warnings` should be handled automatically as well, given that is not part of the standard library

andfoy committed:
CI: do not install pythran from source

andfoy commented:
On a separate note, it seems that the latest commit in Pythran (https://github.com/serge-sans-paille/pythran/commit/b8cd84b5a5ab5222d65781a0194916ad555bf70c) introduced some compilation errors.

andfoy committed:
FIX: remove unnecessary copies on FITPACK routines

andfoy committed:
TST: fix several tests introduced after rebase

andyfaff commented on scipy/optimize/tests/test__differential_evolution.py:
```diff
@@ -849,6 +850,7 @@ def constr_f2(x):
             assert_almost_equal(cv, np.array([[0.0, 0.0, 0.], [2.1, 4.2, 0]]))
             assert cv.shape == (2, 3)
 
+    @pytest.mark.parallel_threads(1)
```
It's not clear to me why this has to be done in a single thread?

(I also have the same question about `test_impossible_constraint`)

andyfaff replied:
Ah ok, because there is something checking that a warning is emitted.

rgommers replied:
Indeed. https://py-free-threading.github.io/porting/#fixing-thread-unsafe-tests was helpful to understand the constructs that can't be used for parallel testing.

rgommers commented:
> We could add `thread_unsafe` as an alias of `parallel_threads(1)`.

Yes please. Could you add that in `pytest-run-parallel`, register the added mark in `scipy/conftest.py`, then replace all the usages in this PR?

> > It feels like a reason= keyword could be helpful here
> 
> We could add that functionality as well, however, it would be purely for informative purposes, since it would not appear on the pytest log trace.

Good point. In that case never mind, a code comment for unusual cases will do.



> We could do that preemptively for instances of functions that use `pytest.warns` indeed, however, we should also look up for instances of `warnings.catch_warnings`, also, I don't know if `np.testing.suppress_warnings` should be handled automatically as well, given that is not part of the standard library

Could you open a new feature request for this in `pytest-run-parallel`? If it's easy and can be done quickly then great, but otherwise let's not have this PR wait for it.

andfoy committed:
TST: fix other tests that use pytest.warns

andfoy committed:
FIX: replace parallel_threads(1) by thread_unsafe

andfoy commented:
Current errors are related to https://github.com/scipy/scipy/issues/21885

andfoy committed:
FIX: remove reference to highs, leftover from rebase

andfoy committed:
TST: mark vode, zvodde and lsoda tests as thread_unsafe

rgommers reviewed (commented):
@andfoy since you just pushed an update, let me submit my review comments that I have so far. I came as far as `optimize`, still need to do the final ~25% of the diff in detail.

I'm happy with the constructs used now. This PR was/is a very heavy lift to review; to make the rest of the thread-safety work easier to land I suggest to go ahead with opening one PR per SciPy submodule now.

There's ~10 comments/questions to look at. Nothing major, just a few tweaks and clarifications I think.

andfoy commented on scipy/ndimage/tests/test_measurements.py:
```diff
@@ -165,7 +165,7 @@ def test_label02(xp):
 
 def test_label03(xp):
-    data = xp.ones([1])
+    data = xp.ones([1], dtype=xp.int64)
```
For some reason, sometimes Cython routines fail with `no matching signature found`, which is related to missing dtype definition, however, I'm not sure why would arrays have different dtypes under parallel conditions

rgommers replied:
Okay, I can reproduce this - but also with the added `dtype=xp.int64`, so that change doesn't fix the issue, perhaps only making it a bit less likely to happen on your machine.

The traceback I am seeing is:
```
__________________________________________________________________ test_label03[numpy] __________________________________________________________________
scipy/ndimage/tests/test_measurements.py:169: in test_label03
    out, n = ndimage.label(data)
        data       = array([1.])
        xp         = <module 'numpy' from '/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/site-packages/numpy/__init__.py'>
scipy/ndimage/_support_alternative_backends.py:37: in wrapper
    result = func(*args, **kwds)
        args       = (array([1.]),)
        delegator  = <function binary_fill_holes_signature at 0x385421dfa00>
        func       = <function label at 0x385421de9c0>
        kwds       = {}
        module_name = 'ndimage'
        xp         = <module 'scipy._lib.array_api_compat.numpy' from '/home/rgommers/code/pixi-dev-scipystack/scipy/scipy/build-nogil-install/lib/python3.13t/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>
scipy/ndimage/_measurements.py:218: in label
    max_label = _ni_label._label(input, structure, output)
        caller_provided_output = False
        ii         = 3
        input      = array([1.])
        need_64bits = False
        output     = array([-889028608], dtype=int32)
        structure  = array([ True,  True,  True])
_ni_label.pyx:200: in _ni_label._label
    ???
_ni_label.pyx:237: in _ni_label._label
    ???
_ni_label.pyx:93: in _ni_label.__pyx_fused_cpdef
    ???
E   TypeError: No matching signature found
```

Both this issue and the one in `_cythonized_array_utils` you're showing above are for code that uses fused types. This is almost certainly a new Cython bug. I suggest to undo the change on this line (so remove `, dtype=xp.int64`) and then open a new issue upstream in Cython for this after creating a cleaner standalone reproducer (this can wait until after this PR is in).

rgommers replied:
To be clear: let's also not add a `thread_unsafe` here for now. Better to spend the time getting it fixed in Cython, rather than trying to work around it and then have to undo the workaround later.

rgommers replied:
The generated C code looks like this:

```C
/* #### Code section: cached_constants ### */
...

  /* "_ni_label.pyx":93
 * # Function specializers
 * ######################################################################
 * def get_nonzero_line(np.ndarray[data_t] a):             # <<<<<<<<<<<<<<
 *     return <Py_intptr_t> fused_nonzero_line[data_t]
 * 
*/
  __pyx_mstate_global->__pyx_tuple[4] = PyTuple_Pack(1, __pyx_mstate_global->__pyx_kp_u_No_matching_signature_found); if (unlikely(!__pyx_mstate_global->__pyx_tuple[4])) __PYX_ERR(0, 93, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_mstate_global->__pyx_tuple[4]);
  __Pyx_GIVEREF(__pyx_mstate_global->__pyx_tuple[4]);
  __pyx_mstate_global->__pyx_tuple[5] = PyTuple_Pack(1, __pyx_mstate_global->__pyx_kp_u_Function_call_with_ambiguous_arg); if (unlikely(!__pyx_mstate_global->__pyx_tuple[5])) __PYX_ERR(0, 93, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_mstate_global->__pyx_tuple[5]);
  __Pyx_GIVEREF(__pyx_mstate_global->__pyx_tuple[5]);

...

  /* "_ni_label.pyx":93
 * # Function specializers
 * ######################################################################
 * def get_nonzero_line(np.ndarray[data_t] a):             # <<<<<<<<<<<<<<
 *     return <Py_intptr_t> fused_nonzero_line[data_t]
 * 
*/

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  { PyObject *__pyx_type, *__pyx_value, *__pyx_tb;
    __Pyx_PyThreadState_declare
    __Pyx_PyThreadState_assign
    __Pyx_ErrFetch(&__pyx_type, &__pyx_value, &__pyx_tb);
    __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_a.rcbuffer->pybuffer);
  __Pyx_ErrRestore(__pyx_type, __pyx_value, __pyx_tb);}
  __Pyx_AddTraceback("_ni_label.get_nonzero_line", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  goto __pyx_L2;
  __pyx_L0:;
  __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_a.rcbuffer->pybuffer);
  __pyx_L2:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_fuse_2__pyx_pw_9_ni_label_13get_nonzero_line(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static PyMethodDef __pyx_fuse_2__pyx_mdef_9_ni_label_13get_nonzero_line = {"__pyx_fuse_2get_nonzero_line", (PyCFunction)(void*)(PyCFunctionWithKeywords)__pyx_fuse_2__pyx_pw_9_ni_label_13get_nonzero_line, METH_VARARGS|METH_KEYWORDS, 0};
static PyObject *__pyx_fuse_2__pyx_pw_9_ni_label_13get_nonzero_line(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  CYTHON_UNUSED PyArrayObject *__pyx_v_a = 0;
  CYTHON_UNUSED Py_ssize_t __pyx_nargs;
  CYTHON_UNUSED PyObject *const *__pyx_kwvalues;
  PyObject* values[1] = {0};
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("get_nonzero_line (wrapper)", 0);
  #if CYTHON_ASSUME_SAFE_SIZE
  __pyx_nargs = PyTuple_GET_SIZE(__pyx_args);
  #else
  __pyx_nargs = PyTuple_Size(__pyx_args); if (unlikely(__pyx_nargs < 0)) return NULL;
  #endif
  __pyx_kwvalues = __Pyx_KwValues_VARARGS(__pyx_args, __pyx_nargs);
  {
    PyObject **__pyx_pyargnames[] = {&__pyx_mstate_global->__pyx_n_u_a,0};
    if (__pyx_kwds) {
      Py_ssize_t kw_args;
      switch (__pyx_nargs) {
        case  1:
        values[0] = __Pyx_Arg_VARARGS(__pyx_args, 0);
        CYTHON_FALLTHROUGH;
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = __Pyx_NumKwargs_VARARGS(__pyx_kwds);
      switch (__pyx_nargs) {
        case  0:
        if (likely((values[0] = __Pyx_GetKwValue_VARARGS(__pyx_kwds, __pyx_kwvalues, __pyx_mstate_global->__pyx_n_u_a)) != 0)) {
          (void)__Pyx_Arg_NewRef_VARARGS(values[0]);
          kw_args--;
        }
        else if (unlikely(PyErr_Occurred())) __PYX_ERR(0, 93, __pyx_L3_error)
        else goto __pyx_L5_argtuple_error;
      }
      if (unlikely(kw_args > 0)) {
        const Py_ssize_t kwd_pos_args = __pyx_nargs;
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_kwvalues, __pyx_pyargnames, 0, values + 0, kwd_pos_args, "get_nonzero_line") < 0)) __PYX_ERR(0, 93, __pyx_L3_error)
      }
    } else if (unlikely(__pyx_nargs != 1)) {
      goto __pyx_L5_argtuple_error;
    } else {
      values[0] = __Pyx_Arg_VARARGS(__pyx_args, 0);
    }
    __pyx_v_a = ((PyArrayObject *)values[0]);
  }
  goto __pyx_L6_skip;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("get_nonzero_line", 1, 1, 1, __pyx_nargs); __PYX_ERR(0, 93, __pyx_L3_error)
  __pyx_L6_skip:;
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  {
    Py_ssize_t __pyx_temp;
    for (__pyx_temp=0; __pyx_temp < (Py_ssize_t)(sizeof(values)/sizeof(values[0])); ++__pyx_temp) {
      __Pyx_Arg_XDECREF_VARARGS(values[__pyx_temp]);
    }
  }
  __Pyx_AddTraceback("_ni_label.get_nonzero_line", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_a), __pyx_mstate_global->__pyx_ptype_5numpy_ndarray, 1, "a", 0))) __PYX_ERR(0, 93, __pyx_L1_error)
  __pyx_r = __pyx_pf_9_ni_label_12get_nonzero_line(__pyx_self, __pyx_v_a);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  {
    Py_ssize_t __pyx_temp;
    for (__pyx_temp=0; __pyx_temp < (Py_ssize_t)(sizeof(values)/sizeof(values[0])); ++__pyx_temp) {
      __Pyx_Arg_XDECREF_VARARGS(values[__pyx_temp]);
    }
  }
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_9_ni_label_12get_nonzero_line(CYTHON_UNUSED PyObject *__pyx_self, CYTHON_UNUSED PyArrayObject *__pyx_v_a) {
  __Pyx_LocalBuf_ND __pyx_pybuffernd_a;
  __Pyx_Buffer __pyx_pybuffer_a;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__pyx_fuse_2get_nonzero_line", 0);
  __pyx_pybuffer_a.pybuffer.buf = NULL;
  __pyx_pybuffer_a.refcount = 0;
  __pyx_pybuffernd_a.data = NULL;
  __pyx_pybuffernd_a.rcbuffer = &__pyx_pybuffer_a;
  {
    __Pyx_BufFmt_StackElem __pyx_stack[1];
    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_a.rcbuffer->pybuffer, (PyObject*)__pyx_v_a, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 93, __pyx_L1_error)
  }
  __pyx_pybuffernd_a.diminfo[0].strides = __pyx_pybuffernd_a.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_a.diminfo[0].shape = __pyx_pybuffernd_a.rcbuffer->pybuffer.shape[0];

  /* "_ni_label.pyx":94
 * ######################################################################
 * def get_nonzero_line(np.ndarray[data_t] a):
 *     return <Py_intptr_t> fused_nonzero_line[data_t]             # <<<<<<<<<<<<<<
 * 
 * def get_read_line(np.ndarray[data_t] a):
*/
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyLong_From_Py_intptr_t(((Py_intptr_t)__pyx_fuse_2__pyx_f_9_ni_label_fused_nonzero_line)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 94, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;
```

And that function definition repeated a bunch of times, with type specifiers like `__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t` inside the `__Pyx_GetBufferAndValidate` call changing. It looks like something isn't thread-safe in that buffer usage.

rgommers commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -1268,6 +1270,7 @@ def test_dc(self, nx, ny, dtype):
         y = signal.resample(x, ny)
         assert_allclose(y, [1] * ny)
 
+    @pytest.mark.thread_unsafe
```
This isn't failing for me locally. The mutable window is copied, so I think it should work? If you see a failure, can you show the traceback?

andfoy replied:
This is another instance of the Cython error

rgommers reviewed (commented):
Last set of comments 😅

andfoy commented on scipy/optimize/tests/test_cobyqa.py:
```diff
@@ -11,6 +11,7 @@
 )
 
 
+@pytest.mark.thread_unsafe
```
Those tests are experiencing numeric mismatches:

```python
E       AssertionError: 
E       Arrays are not equal
E       
E       Mismatched elements: 2 / 2 (100%)
E       Max absolute difference among violations: 5.16253706e-14
E       Max relative difference among violations: 7.74380659e-14
E        ACTUAL: array([4.955356, 0.666667])
E        DESIRED: array([4.955356, 0.666667])

```

rgommers replied:
Okay, I can reproduce. And I think the answer to my question is: yes, `_minimize_cobyqa` needs a lock right now. This change fixes the failures:
```diff
--- a/scipy/optimize/_cobyqa_py.py
+++ b/scipy/optimize/_cobyqa_py.py
@@ -2,6 +2,11 @@ import numpy as np
 
 from ._optimize import _check_unknown_options
 
+from threading import Lock
+
+
+COBYQA_LOCK = Lock()
+
 
 def _minimize_cobyqa(fun, x0, args=(), bounds=None, constraints=(),
                      callback=None, disp=False, maxfev=None, maxiter=None,
@@ -64,4 +69,6 @@ def _minimize_cobyqa(fun, x0, args=(), bounds=None, constraints=(),
         'radius_final': float(final_tr_radius),
         'scale': bool(scale),
     }
-    return minimize(fun, x0, args, bounds, constraints, callback, options)
+
+    with COBYQA_LOCK:
+        return minimize(fun, x0, args, bounds, constraints, callback, options)
```

Can you please verify, and apply that fix and drop the `thread_unsafe` marker in the tests? This can then be followed up with opening an issue on https://github.com/cobyqa/cobyqa/ about the code not being thread-safe for free-threaded CPython. Might also happen on a default `cp313` build, since this is pure Python.

In general, we should be marking tests as `thread_unsafe` if the test code itself is the problem. If the problem is in the functionality under test, it should be made robust. And if that's too hard, adding a comment to the marker and opening an issue so that we can fix it later on.

andfoy commented on scipy/signal/tests/test_cont2discrete.py:
```diff
@@ -13,9 +13,10 @@
 
 
 class TestC2D:
+    @pytest.mark.thread_unsafe
```
Again, another instance of `no matching signature found`

rgommers replied:
How about adding a code comment like:
```
    @pytest.mark.thread_unsafe  # due to Cython fused types, see cython#xxxx
```

If you open a placeholder issue on the Cython issue tracker for now so you can fill in the `xxxx`, and then use the same comment for each test, you can leave the `thread_unsafe` marker. This way we can remove the markers as soon as the Cython bug is fixed. Otherwise we'll be quite puzzled in the future about what is unsafe here.

rgommers replied:
> If you open a placeholder issue on the Cython issue tracker for now so you can fill in the `xxxx`,

@andfoy it would be useful to do this now, and replace `xxxx` with the actual issue number. Should take only a few minutes to open it with a link to the comment thread where we investigated this a bit, and promise them you'll add a clean reproducer within a few days.

andfoy committed:
FIX: Address review comments

andfoy committed:
FIX: Add lock to COBYQA

andfoy committed:
FIX: remove unused imports

rgommers commented on scipy/linalg/tests/test_blas.py:
```diff
@@ -212,6 +212,7 @@ def test_amax(self):
     # XXX: need tests for rot,rotm,rotg,rotmg
 
 
+@pytest.mark.thread_unsafe
```
This change is just removed now, but the tests are both failing and crashing regularly for me now. Why not add the comment?

andfoy committed:
TST: Mark FBLAS tests as thread-unsafe

andfoy committed:
TST: add Cython issue reference to comments

[ci skip]

rgommers committed:
TST: cython: make `test_extending` tests safe for parallel usage

rgommers committed:
BUG: integrate: add locks to lsoda, vode, zvode

This ensures we can actually call `ode` with these solvers
from parallel threads.

`solve_ivp` with LSODA doesn't work, but is well-behaved enough
(it raises a ConcurrencyError). `solve_ivp` has more internal
state and repeatedly calls the underlying solver, so that's
a lot less easy to get to work.

rgommers reviewed (approved):
This now seems happy and all comments are addressed. It makes all the modules that were included for ~98% thread-safe. A few follow-up issues are in progress or to be opened, I'll do that after and link them here. I'll also update the tracking issue (gh-20669) with the details of what this PR covered.

Thanks for the heavy lifting @andfoy!

Note that CI failures were all unrelated, one flaky test fixed in main just now, one timeout issue, the `ifx` build failure is happening in other PRs as well, and the linter complaints are best ignored since there are too many and not in code touched here (only in files touched here).

rgommers commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -64,4 +68,5 @@ def _minimize_cobyqa(fun, x0, args=(), bounds=None, constraints=(),
         'scale': bool(scale),
     }
-    return minimize(fun, x0, args, bounds, constraints, callback, options)
+    with COBYQA_LOCK:
```
@ragonneau this is perhaps of interest to you? I had a quick look through the `cobyqa` package, but it was not easy to tell from a quick browse what wasn't thread-safe.

For context, without this lock, the tests for `cobyqa` within SciPy show that numerically incorrect solutions are returned from `cobyqa` when calling it from multiple threads in parallel (as done by `pytest-run-parallel` in this PR).

ragonneau replied:
Hi @rgommers,

I am currently extremely busy, but I will have a look in the future. Do you mean that the module is not deterministic (in the sense that two different solutions are returned when submitting the exact same inputs) when run in parallel? I actually tested this a lot in the past, but I might have missed something...

Cheers,
Tom.

rgommers replied:
Yes indeed, that seems to be the case. `cobyqa` is pure Python, so it's not impossible that it's a bug in CPython's free-threading support (it is specific to free-threading, so that hasn't really been tested before).

Anyway, this is not urgent at all, so no worries. 

I'll include the failures for reference:

<details>

```python
scipy/optimize/tests/test_cobyqa.py F...F..F                                               [100%]

============================================ FAILURES ============================================
________________________________ TestCOBYQA.test_minimize_simple _________________________________
scipy/optimize/tests/test_cobyqa.py:80: in test_minimize_simple
    assert_equal(sol.x, sol_new.x)
        Callback   = <class 'scipy.optimize.tests.test_cobyqa.TestCOBYQA.test_minimize_simple.<locals>.Callback'>
        CallbackNewSyntax = <class 'scipy.optimize.tests.test_cobyqa.TestCOBYQA.test_minimize_simple.<locals>.CallbackNewSyntax'>
        callback   = <scipy.optimize.tests.test_cobyqa.TestCOBYQA.test_minimize_simple.<locals>.Callback object at 0x4c9f40e0190>
        callback_new_syntax = <scipy.optimize.tests.test_cobyqa.TestCOBYQA.test_minimize_simple.<locals>.CallbackNewSyntax object at 0x4c9f40e0310>
        constraints = <scipy.optimize._constraints.NonlinearConstraint object at 0x4c9f40f0150>
        self       = <scipy.optimize.tests.test_cobyqa.TestCOBYQA object at 0x4c9e1aec810>
        sol        =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.85185186080888
       x: [ 4.955e+00  6.667e-01]
     nit: 42
   maxcv: 8.95702001457721e-09
    nfev: 30
        sol_new    =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.851851860808875
       x: [ 4.955e+00  6.667e-01]
     nit: 42
   maxcv: 8.957012909149853e-09
    nfev: 30
        solution   = [np.float64(4.955356249106169), 0.6666666666666666]
        x0         = [4.95, 0.66]
../../../../../.pixi/envs/free-threading/lib/python3.13t/site-packages/numpy/_utils/__init__.py:85: in wrapper
    return fun(*args, **kwargs)
        args       = (array([4.95535626, 0.66666658]), array([4.95535626, 0.66666658]), '', True)
        dep_version = '2.0.0'
        fun        = <function assert_array_equal at 0x4c9e2ffc8a0>
        kwargs     = {'strict': False}
        new_name   = 'desired'
        new_names  = ['actual', 'desired']
        old_name   = 'y'
        old_names  = ['x', 'y']
../../../../../.pixi/envs/free-threading/lib/python3.13t/contextlib.py:85: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Arrays are not equal
E   
E   Mismatched elements: 2 / 2 (100%)
E   Max absolute difference among violations: 6.28386232e-14
E   Max relative difference among violations: 9.42579469e-14
E    ACTUAL: array([4.955356, 0.666667])
E    DESIRED: array([4.955356, 0.666667])
        args       = (<built-in function eq>, array([4.95535626, 0.66666658]), array([4.95535626, 0.66666658]))
        func       = <function assert_array_compare at 0x4c9e2ffc6c0>
        kwds       = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}
        self       = <contextlib._GeneratorContextManager object at 0x4c9e07e64b0>
_________________________________ TestCOBYQA.test_minimize_array _________________________________
scipy/optimize/tests/test_cobyqa.py:187: in test_minimize_array
    assert_equal(sol.x, sol_array.x)
        bounds     = Bounds(array([4.5, 0.6]), array([5. , 0.7]))
        constraints = <scipy.optimize._constraints.NonlinearConstraint object at 0x4c9f21800b0>
        dim        = 0
        fun_array  = <function TestCOBYQA.test_minimize_array.<locals>.fun_array at 0x4c9f2060500>
        self       = <scipy.optimize.tests.test_cobyqa.TestCOBYQA object at 0x4c9e1aecf90>
        sol        =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.851851852351608
       x: [ 4.955e+00  6.667e-01]
     nit: 36
   maxcv: 4.997531277695089e-10
    nfev: 25
        sol_array  =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.851851852351484
       x: [ 4.955e+00  6.667e-01]
     nit: 36
   maxcv: 4.996287827907508e-10
    nfev: 25
../../../../../.pixi/envs/free-threading/lib/python3.13t/site-packages/numpy/_utils/__init__.py:85: in wrapper
    return fun(*args, **kwargs)
        args       = (array([4.95535624, 0.66666673]), array([4.95535624, 0.66666672]), '', True)
        dep_version = '2.0.0'
        fun        = <function assert_array_equal at 0x4c9e2ffc8a0>
        kwargs     = {'strict': False}
        new_name   = 'desired'
        new_names  = ['actual', 'desired']
        old_name   = 'y'
        old_names  = ['x', 'y']
../../../../../.pixi/envs/free-threading/lib/python3.13t/contextlib.py:85: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Arrays are not equal
E   
E   Mismatched elements: 2 / 2 (100%)
E   Max absolute difference among violations: 2.81543933e-09
E   Max relative difference among violations: 4.22315862e-09
E    ACTUAL: array([4.955356, 0.666667])
E    DESIRED: array([4.955356, 0.666667])
        args       = (<built-in function eq>, array([4.95535624, 0.66666673]), array([4.95535624, 0.66666672]))
        func       = <function assert_array_compare at 0x4c9e2ffc6c0>
        kwds       = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}
        self       = <contextlib._GeneratorContextManager object at 0x4c9e07e64b0>
_______________________________ TestCOBYQA.test_minimize_f_target ________________________________
scipy/optimize/tests/test_cobyqa.py:251: in test_minimize_f_target
    assert sol.nfev <= sol_ref.nfev, sol
E   AssertionError:  message: The target objective function value has been reached
E      success: True
E       status: 1
E          fun: 24.851851851938974
E            x: [ 4.955e+00  6.667e-01]
E          nit: 37
E        maxcv: 8.712319754522468e-11
E         nfev: 32
E   assert 32 <= 30
E    +  where 32 =  message: The target objective function value has been reached\n success: True\n  status: 1\n     fun: 24.851851851938974\n       x: [ 4.955e+00  6.667e-01]\n     nit: 37\n   maxcv: 8.712319754522468e-11\n    nfev: 32.nfev
E    +  and   30 =  message: The lower bound for the trust-region radius has been reached\n success: True\n  status: 0\n     fun: 24.851851860808875\n       x: [ 4.955e+00  6.667e-01]\n     nit: 42\n   maxcv: 8.957016461863532e-09\n    nfev: 30.nfev
        constraints = <scipy.optimize._constraints.NonlinearConstraint object at 0x4c9f21803d0>
        options    = {'f_target': np.float64(24.851851860808875), 'maxfev': 100}
        self       = <scipy.optimize.tests.test_cobyqa.TestCOBYQA object at 0x4c9e0628c10>
        sol        =  message: The target objective function value has been reached
 success: True
  status: 1
     fun: 24.851851851938974
       x: [ 4.955e+00  6.667e-01]
     nit: 37
   maxcv: 8.712319754522468e-11
    nfev: 32
        sol_ref    =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.851851860808875
       x: [ 4.955e+00  6.667e-01]
     nit: 42
   maxcv: 8.957016461863532e-09
    nfev: 30
==================================== short test summary info =====================================
FAILED scipy/optimize/tests/test_cobyqa.py::TestCOBYQA::test_minimize_simple - AssertionError: 
FAILED scipy/optimize/tests/test_cobyqa.py::TestCOBYQA::test_minimize_array - AssertionError: 
FAILED scipy/optimize/tests/test_cobyqa.py::TestCOBYQA::test_minimize_f_target - AssertionError:  message: The target objective function value has been reached
```

</details>

The first two failures still give solutions, just with slightly lower precision (that may be because we're running the exact same problem in parallel though - if there's cross-talk between threads, running different problems in parallel may yield larger errors). The last failure looks a little more serious.

mdhaber commented on scipy/integrate/_tanhsinh.py:
```diff
@@ -373,7 +373,9 @@ def _tanhsinh(f, a, b, *, args=(), log=False, maxfun=None, maxlevel=None,
         xr0=xr0, fr0=fr0, wr0=wr0, xl0=xl0, fl0=fl0, wl0=wl0, d4=d4,  # err est
-        ainf=ainf, binf=binf, abinf=abinf, a0=xp.reshape(a0, (-1, 1)))  # transforms
+        ainf=ainf, binf=binf, abinf=abinf, a0=xp.reshape(a0, (-1, 1)),  # transforms
+        pc_xjc=None, pc_wj=None, pc_indices=[0], pc_h0=None)  # pair cache
```
Please let me know so I can avoid the pattern in the future. I ran into this because it introduced a subtle bug (which I'll fix before the release). 

I understand that global caches can be problematic, but there was a reason for it, so I wonder whether there is an acceptable pattern for maintaining state between calls to a function?

rgommers replied:
If it's easy to explain what the bug is, can you explain? If not, happy to look at the PR when it's ready.

A global cache needs _something_ to make it safe in general. Caches should be protected by a lock or some other way to make them safe for reading/writing in parallel threads. Or made thread-local. Never use a regular global variable for anything unless it's completely immutable.

I don't entirely remember the context since it's 3 weeks ago I looked at it, but this doesn't look like an expensive function call that needs a cache but more like an iterative algorithm that needs the state of the previous iteration. In general, it's cleanest to pass that into the function.

mdhaber replied:
> If it's easy to explain what the bug is, can you explain?

Arrays in `work` are subject to being "compressed" - elements corresponding with calculations that have already converged may be eliminated so no further computation is done with them. In some situtations, `pc_xjc`/`pc_wj` were being compressed, but that's not what we want. The simple fix is to protect them by putting them in another container (the `_RichResult` used in `derivative`, for instance). Admittedly, this can also be avoided in the future by separating compressible and incompressible parts of `work` more explicitly, which is something I've planned to do when things slow down a bit.

> this doesn't look like an expensive function call that needs a cache but more like an iterative algorithm that needs the state of the previous iteration. In general, it's cleanest to pass that into the function.

In these cases, the caches were for things like abscissae and weights (e.g. like the outputs of [`roots_legendre`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.roots_legendre.html#scipy.special.roots_legendre)). In this case, the abscissae/weights are mostly independent of the problem being solved... the list of them just grows if the algorithm requires more iterations. The intent was to generate as many as needed the first time the function runs, rather than *every* time, and only ever expand the cache after that. This can be a significant time savings when, say, quadrature is being done within an iterative root-finder (e.g. `ppf` of a distribution that computes the CDF by integrating the PDF). 

The cache *can* change in some conditions. In `tanhsinh`, `h0` is dependent on the dtype, so cache needs to change if different dtype (or array type) is used. In `derivative`, I think they only change if the `step_factor` option is changed. I suppose the user could be working with different dtypes or `step_factor`s in different threads, so it's good to fix, but I'd like to find a way that doesn't reduce the performance in the typical case.

rgommers replied:
Thanks. It sounds like you want something like [`functools.lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache) per dtype then.

For different `step_factor` I'm not sure it makes sense to cache, because the cache size could explore if a user tries different step sizes in a loop (unlike dtypes it's not a fixed set).

mdhaber replied:
Just to be clear, it's factors by which the step size is reduced, not the step size itself. This is adaptive, so the user shouldn't need to choose either, but if they do want to try different *initial* step sizes, they can do so without the cache changing at all. (In fact, the code is vectorized w.r.t. initial step, so they can try many in a single call).

When the factor changes, the old cache is replaced. So the cache wouldn't explode if the user loops over step factors.

A problem would (only?) occur if different threads were trying different step factors at the same time. 

I will try `lru_cache`. IIRC that is not thread safe by default, and it requires that arguments are hashable, but it can probably be made to work.

rgommers replied:
> I will try `lru_cache`. IIRC that is not thread safe by default,

Its docs say: _"The cache is threadsafe so that the wrapped function can be used in multiple threads. This means that the underlying data structure will remain coherent during concurrent updates."_

"concurrent updates" (i.e. 2 threads writing at once, possibly the same value, or 1 read and 1 write at once) is normally what thread safety means. Your description of what you want is not that though, and that's because your design sounds like it's not really a cache. A cache has the property that for the same inputs, the outputs are always identical. If that's not the case, you have a bigger issue. If `f(x, y)` returns `42` with the default value of some unknown hidden parameter `z`, and if the user tweaks `z` and the answer becomes `43`, then it sounds like that's inherently not cache-able. `z` must then either become an explicit input parameter so the cache knows about it, or you move the cache so that `z` can no longer influence the result.

> A problem would (only?) occur if different threads were trying different step factors at the same time.

Indeed. It's a bug on any Python version, not only with free-threading. A user can always use the `threading` module.

rgommers replied:
There are a number of explicit tests for concurrency in the code base, named `test_concurrency` or `test_threading` typically. Adopting the test pattern in one of those may help here.

mdhaber commented on scipy/differentiate/tests/test_differentiate.py:
```diff
@@ -54,17 +55,19 @@ def test_vectorization(self, order, shape, xp):
+            state['feval'] += 1 if (x.size == n or x.ndim <=1) else x.shape[-1]
             return self.f(x, *args, **kwargs)
-        f.nit = -1
-        f.feval = 0
```
Is there an automatic check now for this sort of thing to prevent this sort of thing in the future?

rgommers replied:
There will be once we're done with all of this "make things thread-safe" set of PRs. Namely a CI run with [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel/).
